libs/server-sdk/src/data_components/serialization_adapters/json_pull_source.hpp

```diff
@@ -0,0 +1,72 @@
+#pragma once
+
+#include "../../data_interfaces/source/ipull_source.hpp"
+#include "../kinds/kinds.hpp"
+
+#include <launchdarkly/server_side/data_interfaces/sources/iserialized_pull_source.hpp>
+
+namespace launchdarkly::server_side::data_components {
+
+class JsonSource final : public data_interfaces::IPullSource {
+   public:
+    explicit JsonSource(
+        data_interfaces::ISerializedDataPullSource& json_source);
+
+    [[nodiscard]] Single<data_model::FlagDescriptor> GetFlag(
+        std::string const& key) const override;
+
+    [[nodiscard]] Single<data_model::SegmentDescriptor> GetSegment(
+        std::string const& key) const override;
+
+    [[nodiscard]] Collection<data_model::FlagDescriptor> AllFlags()
+        const override;
+
+    [[nodiscard]] Collection<data_model::SegmentDescriptor> AllSegments()
+        const override;
+
+    [[nodiscard]] std::string const& Identity() const override;
+
+   private:
+    template <typename DataModel, typename DataKind>
+    Single<data_model::ItemDescriptor<DataModel>> Deserialize(
+        DataKind const& kind,
+        std::string const& key) const {
+        auto result = source_.Get(kind, key);
+
+        if (!result) {
+            /* the actual fetch failed */
+            return tl::make_unexpected(result.error().message);
+        }
+
+        if (!result->serializedItem) {
+            /* the fetch succeeded, but the item wasn't found */
+            return std::nullopt;
+        }
+
+        auto const boost_json_val = boost::json::parse(*result->serializedItem);
+        auto flag = boost::json::value_to<
+            tl::expected<std::optional<DataModel>, JsonError>>(boost_json_val);
+
+        if (!flag) {
+            /* flag couldn't be deserialized from the JSON string */
+            return tl::make_unexpected(ErrorToString(flag.error()));
+        }
+
+        std::optional<DataModel> maybe_flag = flag->value();
+
+        if (!maybe_flag) {
+            /* JSON was valid, but the value is 'null'
+             * TODO: will this ever happen?
+             */
+            return tl::make_unexpected("data was null");
```

Comment:
This is the case where we get back a serialized item, but then upon deserializing the JSON, it's actually `null`.

The deserializer implementation (called from `value_to`) was designed to return an `std::optional` so the caller could distinguish "got a valid object" vs omitted where it is important to do so.

In this case, I don't think we'd ever store `null` in the persistent store so don't think it would come up in practice.